File: src/domain/models/doc_health.py
"""Domain models for document health checking.

DocHealthStatus enumerates possible doc states.
DocRegistryEntry records a doc to track with its review interval.
DocStatus captures the health of a single document.
DocHealthReport aggregates statuses into a summary.
create_doc_status is a factory that auto-calculates status and days_since.
"""

from __future__ import annotations

import enum
from dataclasses import dataclass
from datetime import date

from src.domain.models.errors import InvariantViolationError


class DocHealthStatus(enum.Enum):
    """Possible health states for a tracked document."""

    OK = "ok"
    STALE = "stale"
    MISSING = "missing"
    NO_FRONTMATTER = "no_frontmatter"


@dataclass(frozen=True)
class DocRegistryEntry:
    """A document registered for health tracking.

    Attributes:
        path: Relative path to the document from project root.
        owner: Optional owner or team responsible for the document.
        review_interval_days: Maximum days between reviews before staleness.
    """

    path: str
    owner: str | None = None
    review_interval_days: int = 30

    def __post_init__(self) -> None:
        if self.review_interval_days <= 0:
            msg = (
                f"review_interval_days must be positive, got {self.review_interval_days}"
            )
            raise InvariantViolationError(msg)


@dataclass(frozen=True)
class DocStatus:
    """Health status of a single document.

    Attributes:
        path: Relative path to the document.
        status: Current health status.
        last_reviewed: Date the document was last reviewed, if known.
        days_since: Days since last review, if known.
        review_interval_days: The review interval for this document.
        owner: Optional owner of this document.
    """

    path: str
    status: DocHealthStatus
    last_reviewed: date | None = None
    days_since: int | None = None
    review_interval_days: int = 30
    owner: str | None = None


@dataclass(frozen=True)
class DocHealthReport:
    """Aggregate report of document health statuses.

    Attributes:
        statuses: Tuple of individual document statuses.
    """

    statuses: tuple[DocStatus, ...]

    @property
    def issue_count(self) -> int:
        """Count of documents with non-OK status."""
        return sum(1 for s in self.statuses if s.status != DocHealthStatus.OK)

    @property
    def total_checked(self) -> int:
        """Total number of documents checked."""
        return len(self.statuses)

    @property
    def has_issues(self) -> bool:
        """Whether any document has a non-OK status."""
        return self.issue_count > 0


def create_doc_status(
    path: str,
    exists: bool,
    last_reviewed: date | None,
    review_interval_days: int = 30,
    owner: str | None = None,
    today: date | None = None,
) -> DocStatus:
    """Create DocStatus with auto-calculated status and days_since.

    Args:
        path: Relative path to the document.
        exists: Whether the file exists on disk.
        last_reviewed: Date of last review, or None if unknown.
        review_interval_days: Maximum days between reviews.
        owner: Optional owner of this document.

    Returns:
        A DocStatus with the appropriate status and computed days_since.
    """
    if not exists:
        return DocStatus(
            path=path,
            status=DocHealthStatus.MISSING,
            review_interval_days=review_interval_days,
            owner=owner,
        )
    if last_reviewed is None:
        return DocStatus(
            path=path,
            status=DocHealthStatus.NO_FRONTMATTER,
            review_interval_days=review_interval_days,
            owner=owner,
        )
    effective_today = today if today is not None else date.today()
    days = (effective_today - last_reviewed).days
    status = DocHealthStatus.OK if days <= review_interval_days else DocHealthStatus.STALE
    return DocStatus(
        path=path,
        status=status,
        last_reviewed=last_reviewed,
        days_since=days,
        review_interval_days=review_interval_days,
        owner=owner,
    )

File: internal/dochealth/domain/doc_health.go
// Package domain provides the DocHealth bounded context's core domain model.
// It contains value objects for document health checking: statuses, registry entries,
// broken links, doc statuses, health reports, and review results.
package domain

import (
	"fmt"
	"time"

	domainerrors "github.com/alty-cli/alty/internal/shared/domain/errors"
)

// DocHealthStatus enumerates possible document health states.
type DocHealthStatus string

// Document health status constants.
const (
	DocHealthOK            DocHealthStatus = "ok"
	DocHealthStale         DocHealthStatus = "stale"
	DocHealthMissing       DocHealthStatus = "missing"
	DocHealthNoFrontmatter DocHealthStatus = "no_frontmatter"
)

// BrokenLink is a broken internal markdown link found in a document.
type BrokenLink struct {
	linkText   string
	target     string
	reason     string
	lineNumber int
}

// NewBrokenLink creates a validated BrokenLink value object.
func NewBrokenLink(lineNumber int, linkText, target, reason string) (BrokenLink, error) {
	if lineNumber < 1 {
		return BrokenLink{}, fmt.Errorf("line_number must be >= 1, got %d: %w",
			lineNumber, domainerrors.ErrInvariantViolation)
	}
	return BrokenLink{
		lineNumber: lineNumber,
		linkText:   linkText,
		target:     target,
		reason:     reason,
	}, nil
}

// LineNumber returns the line number where the broken link was found.
func (bl BrokenLink) LineNumber() int { return bl.lineNumber }

// LinkText returns the link text.
func (bl BrokenLink) LinkText() string { return bl.linkText }

// Target returns the link target.
func (bl BrokenLink) Target() string { return bl.target }

// Reason returns why the link is broken.
func (bl BrokenLink) Reason() string { return bl.reason }

// DocRegistryEntry is a document registered for health tracking.
type DocRegistryEntry struct {
	path               string
	owner              string
	reviewIntervalDays int
}

// NewDocRegistryEntry creates a validated DocRegistryEntry.
func NewDocRegistryEntry(path, owner string, reviewIntervalDays int) (DocRegistryEntry, error) {
	if reviewIntervalDays <= 0 {
		return DocRegistryEntry{}, fmt.Errorf(
			"review_interval_days must be positive, got %d: %w",
			reviewIntervalDays, domainerrors.ErrInvariantViolation)
	}
	return DocRegistryEntry{
		path:               path,
		owner:              owner,
		reviewIntervalDays: reviewIntervalDays,
	}, nil
}

// Path returns the document path.
func (e DocRegistryEntry) Path() string { return e.path }

// Owner returns the document owner.
func (e DocRegistryEntry) Owner() string { return e.owner }

// ReviewIntervalDays returns the review interval in days.
func (e DocRegistryEntry) ReviewIntervalDays() int { return e.reviewIntervalDays }

// DocStatus captures the health of a single document.
type DocStatus struct {
	path               string
	status             DocHealthStatus
	lastReviewed       *time.Time
	daysSince          *int
	owner              string
	brokenLinks        []BrokenLink
	reviewIntervalDays int
}

// NewDocStatus creates a DocStatus value object directly.
func NewDocStatus(
	path string,
	status DocHealthStatus,
	lastReviewed *time.Time,
	daysSince *int,
	reviewIntervalDays int,
	owner string,
	brokenLinks []BrokenLink,
) DocStatus {
	bl := make([]BrokenLink, len(brokenLinks))
	copy(bl, brokenLinks)
	return DocStatus{
		path:               path,
		status:             status,
		lastReviewed:       lastReviewed,
		daysSince:          daysSince,
		reviewIntervalDays: reviewIntervalDays,
		owner:              owner,
		brokenLinks:        bl,
	}
}

// Path returns the document path.
func (s DocStatus) Path() string { return s.path }

// Status returns the document health status.
func (s DocStatus) Status() DocHealthStatus { return s.status }

// LastReviewed returns the last review timestamp.
func (s DocStatus) LastReviewed() *time.Time { return s.lastReviewed }

// DaysSince returns the number of days since last review.
func (s DocStatus) DaysSince() *int { return s.daysSince }

// ReviewIntervalDays returns the review interval in days.
func (s DocStatus) ReviewIntervalDays() int { return s.reviewIntervalDays }

// Owner returns the document owner.
func (s DocStatus) Owner() string { return s.owner }

// BrokenLinks returns a defensive copy.
func (s DocStatus) BrokenLinks() []BrokenLink {
	out := make([]BrokenLink, len(s.brokenLinks))
	copy(out, s.brokenLinks)
	return out
}

// CreateDocStatus is a factory that auto-calculates status and days_since.
func CreateDocStatus(
	path string,
	exists bool,
	lastReviewed *time.Time,
	reviewIntervalDays int,
	owner string,
	today *time.Time,
	brokenLinks []BrokenLink,
) DocStatus {
	bl := make([]BrokenLink, len(brokenLinks))
	copy(bl, brokenLinks)

	if !exists {
		return DocStatus{
			path:               path,
			status:             DocHealthMissing,
			reviewIntervalDays: reviewIntervalDays,
			owner:              owner,
			brokenLinks:        bl,
		}
	}
	if lastReviewed == nil {
		return DocStatus{
			path:               path,
			status:             DocHealthNoFrontmatter,
			reviewIntervalDays: reviewIntervalDays,
			owner:              owner,
			brokenLinks:        bl,
		}
	}

	effectiveToday := time.Now().Truncate(24 * time.Hour)
	if today != nil {
		effectiveToday = *today
	}
	days := int(effectiveToday.Sub(*lastReviewed).Hours() / 24)

	status := DocHealthOK
	if days > reviewIntervalDays {
		status = DocHealthStale
	}

	lr := *lastReviewed
	return DocStatus{
		path:               path,
		status:             status,
		lastReviewed:       &lr,
		daysSince:          &days,
		reviewIntervalDays: reviewIntervalDays,
		owner:              owner,
		brokenLinks:        bl,
	}
}

// DocHealthReport aggregates document health statuses into a summary.
type DocHealthReport struct {
	statuses []DocStatus
}

// NewDocHealthReport creates a DocHealthReport.
func NewDocHealthReport(statuses []DocStatus) DocHealthReport {
	s := make([]DocStatus, len(statuses))
	copy(s, statuses)
	return DocHealthReport{statuses: s}
}

// Statuses returns a defensive copy of all statuses.
func (r DocHealthReport) Statuses() []DocStatus {
	out := make([]DocStatus, len(r.statuses))
	copy(out, r.statuses)
	return out
}

// IssueCount returns the count of documents with non-OK status or broken links.
func (r DocHealthReport) IssueCount() int {
	count := 0
	for _, s := range r.statuses {
		if s.status != DocHealthOK || len(s.brokenLinks) > 0 {
			count++
		}
	}
	return count
}

// TotalChecked returns the total number of documents checked.
func (r DocHealthReport) TotalChecked() int {
	return len(r.statuses)
}

// HasIssues returns whether any document has a non-OK status or broken links.
func (r DocHealthReport) HasIssues() bool {
	return r.IssueCount() > 0
}

// DocReviewResult is the result of marking a document as reviewed.
type DocReviewResult struct {
	newDate time.Time
	path    string
}

// NewDocReviewResult creates a DocReviewResult.
func NewDocReviewResult(path string, newDate time.Time) DocReviewResult {
	return DocReviewResult{path: path, newDate: newDate}
}

// Path returns the document path.
func (r DocReviewResult) Path() string { return r.path }

// NewDate returns the new review date.
func (r DocReviewResult) NewDate() time.Time { return r.newDate }
